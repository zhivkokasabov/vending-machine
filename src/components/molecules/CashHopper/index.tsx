import React from "react";

import { Coin } from "components/atoms/Coin";
import { useCashHopper } from "providers/CashHopper.provider";

export const CashHopper = () => {
  const { stashedCoins } = useCashHopper();

  const stackDefaultValue: Record<string, number> = {};
  const groupedCoins = stashedCoins.reduce((acc, curr) => {
    const roundedKey = curr.toFixed(2);

    acc[roundedKey] = acc[roundedKey] ? acc[roundedKey] + 1 : 1;

    return acc;
  }, stackDefaultValue);
  // might sort them eventually, we'll see
  const stacks = Object.entries(groupedCoins);

  return (
    <div className="flex gap-2 pt-4 w-full max-w-full flex-wrap" data-testid="cash-hopper">
      {stacks.map(([denomination, stackSize]) => {
        return <div key={denomination} className="relative w-[3rem] h-[3rem]">
          {
            Array.from({ length: stackSize > 5 ? 5 : stackSize }, (_, index) => {
              return <Coin
                key={index}
                denomination={denomination}
                // generate for: top-[-3px] top-[-6px] top-[-9px] top-[-12px] top-[-15px]
                className={`absolute left-0 top-[${index * -3}px]`}
              />;
            })
          }
          {
            stackSize < 5
              ? (null)
              : (
                <div className="absolute bottom-[0.5rem] right-[-1px] rounded-full bg-white border border-black text-center w-4 h-4 text-xs">{stackSize}</div>
              )
          }
        </div>
      })}
    </div>
  )
}