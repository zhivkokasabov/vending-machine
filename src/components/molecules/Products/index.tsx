import Image from 'next/image';

import { useProducts } from "providers/Products.provider";

export const Products = () => {
  const { products } = useProducts();

  return <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 lg:grid-cols-6 grow gap-2 gap-y-8">
    {
      products.map((product) => (
        <div key={product.id} className="flex flex-col" data-testid="product">
          <div className="relative text-center grow pt-[50%]">
            <Image
              src={product.image}
              alt={product.name}
              style={{ objectFit: "cover" }}
              fill
              priority
              sizes="(max-width: 640px) 100vw, (max-width: 769px) 50vw, (max-width: 1200px) 25vw"
            />
            {
              product.stock > 0
                ? null
                : (
                  <div className="absolute left-1/2 top-1/2 bg-white shadow -translate-x-1/2 -translate-y-1/2 text-nowrap px-2 py-1 text-sm">
                    Out of stock
                  </div>
                )
            }
          </div>
          <div className="flex justify-center">
            <span className="py-1 px-2 text-sm bg-white">
              { product.code }
            </span>
            <span className="py-1 px-2 text-sm bg-black text-white">
              { product.price.toFixed(2) }
            </span>
          </div>
        </div>
      ))
    }
  </div>
};
