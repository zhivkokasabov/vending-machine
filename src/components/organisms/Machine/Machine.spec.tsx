import { Machine } from ".";

import { ByRoleMatcher, fireEvent, render, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { CoinManagerProvider } from "providers/CoinManager.provider";
import { CashHopperProvider } from "providers/CashHopper.provider";
import { CodeProvider } from "providers/Code.provider";
import { ProductsProvider } from "providers/Products.provider";
import { MessageProvider } from "providers/Message.provider";
import { IProduct } from "interfaces/Product.interface";

import { TooltipProvider } from "components/ui/tooltip";

jest.mock("utils/generateRandomDenomination.util", () => ({
  generateRandomDenomination: () => 2.05
}));

describe("Machine tests", () => {
  const products: IProduct[] = [
    {
      id: 1,
      name: "Soda Can",
      stock: 1,
      image: "https://picsum.photos/600/300?text=Soda+Can",
      code: "001",
      price: 1.25
    },
    {
      id: 2,
      name: "Candy Bar",
      stock: 0,
      image: "https://picsum.photos/600/300?text=Candy+Bar",
      code: "002",
      price: 0.75
    }
  ];

  const renderComponent = () => render(
    <TooltipProvider>
      <CoinManagerProvider>
        <CashHopperProvider>
          <CodeProvider>
            <ProductsProvider products={products}>
              <MessageProvider>
                <Machine />
              </MessageProvider>
            </ProductsProvider>
          </CodeProvider>
        </CashHopperProvider>
      </CoinManagerProvider>
    </TooltipProvider>
  );

  const getButtons = (getAllByRole: (role: ByRoleMatcher) => HTMLElement[]) => {
    const buttons = getAllByRole("button");
    const denominations = buttons.slice(0, 8);
    const controls = [buttons[19]].concat(buttons.slice(9, 18));

    return {
      denominations,
      controls,
      random: buttons[8],
      purchase: buttons[21],
      cancel: buttons[18],
      cashOut: buttons[20]
    };
  }

  it("should render", () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should sum inserted coins", () => {
    const { getAllByRole, queryByText } = renderComponent();

    const { denominations } = getButtons(getAllByRole);

    fireEvent.click(denominations[2]);
    fireEvent.click(denominations[2]);
    fireEvent.click(denominations[1]);
    fireEvent.click(denominations[5]);
    fireEvent.click(denominations[3]);
    fireEvent.click(denominations[6]);

    expect(queryByText("1.72")).not.toBeNull();
  });

  it("should send invalid denominations to the cashHopper", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { random } = getButtons(getAllByRole);
    const cashHopper = within(getByTestId("cash-hopper"));
    const display = within(getByTestId("display"));

    await userEvent.click(random);

    expect(cashHopper.queryByText("2.05")).not.toBeNull();
    expect(display.queryByText("2.05")).toBeNull();
  });

  it("should append control characters to code", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { controls } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(controls[3]);
    await userEvent.click(controls[7]);
    await userEvent.click(controls[4]);
    await userEvent.click(controls[0]);

    expect(display.queryByText("3740")).not.toBeNull();
  });

  it("should clear code on cancel click", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { controls, cancel } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(controls[2]);
    await userEvent.click(controls[6]);
    await userEvent.click(controls[3]);
    await userEvent.click(controls[9]);
    await userEvent.click(cancel);

    expect(display.queryByText("3740")).toBeNull();
  });

  it("should reset total on cashOut", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { denominations, cashOut } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(denominations[6]);
    await userEvent.click(denominations[6]);
    await userEvent.click(cashOut);

    expect(display.queryByText("0")).not.toBeNull();
  });

  it("should return change on cashOut", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { denominations, cashOut } = getButtons(getAllByRole);
    const cashHopper = within(getByTestId("cash-hopper"));

    await userEvent.click(denominations[1]);
    await userEvent.click(denominations[1]);
    await userEvent.click(denominations[2]);
    await userEvent.click(cashOut);

    expect(cashHopper.queryByText("0.05")).not.toBeNull();
    expect(cashHopper.queryAllByText("0.02")).toHaveLength(2);
  });

  it("should show error message if invalid code", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { purchase, controls } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(controls[1]);
    await userEvent.click(controls[1]);
    await userEvent.click(controls[2]);
    await userEvent.click(purchase);

    expect(display.queryByText("Invalid code!")).not.toBeNull();
  });

  it("should show error message if insufficient funds", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { purchase, controls } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(controls[0]);
    await userEvent.click(controls[0]);
    await userEvent.click(controls[1]);

    await userEvent.click(purchase);

    expect(display.queryByText("Insufficient funds!")).not.toBeNull();
  });

  it("should show error message if product is out of stock", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { purchase, controls } = getButtons(getAllByRole);
    const display = within(getByTestId("display"));

    await userEvent.click(controls[0]);
    await userEvent.click(controls[0]);
    await userEvent.click(controls[2]);

    await userEvent.click(purchase);

    expect(display.queryByText("Product is out of stock!")).not.toBeNull();
  });

  it("should subtract from product stock on purchase", async () => {
    const { getAllByRole, getAllByTestId } = renderComponent();

    const { purchase, controls, denominations } = getButtons(getAllByRole);
    const product = within(getAllByTestId("product")[0]);

    await userEvent.click(controls[0]);
    await userEvent.click(controls[0]);
    await userEvent.click(controls[1]);
    
    await userEvent.click(denominations[7]);

    await userEvent.click(purchase);

    expect(product.queryByText("Out of stock")).not.toBeNull();
  });

  it("should return change on purchase", async () => {
    const { getAllByRole, getByTestId } = renderComponent();

    const { purchase, controls, denominations } = getButtons(getAllByRole);
    const cashHopper = within(getByTestId("cash-hopper"));

    await userEvent.click(controls[0]);
    await userEvent.click(controls[0]);
    await userEvent.click(controls[1]);
    
    await userEvent.click(denominations[7]);

    await userEvent.click(purchase);

    expect(cashHopper.queryByText("0.50")).not.toBeNull();
    expect(cashHopper.queryByText("0.20")).not.toBeNull();
    expect(cashHopper.queryByText("0.05")).not.toBeNull();
  });
});
