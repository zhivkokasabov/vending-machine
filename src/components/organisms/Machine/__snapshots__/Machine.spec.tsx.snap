// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Machine tests should render 1`] = `
<DocumentFragment>
  <div
    class="flex gap-4 grow mb-4"
  >
    <div
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 lg:grid-cols-6 grow gap-2 gap-y-8"
    >
      <div
        class="flex flex-col"
        data-testid="product"
      >
        <div
          class="relative text-center grow pt-[50%]"
        >
          <img
            alt="Soda Can"
            data-nimg="fill"
            decoding="async"
            sizes="(max-width: 640px) 100vw, (max-width: 769px) 50vw, (max-width: 1200px) 25vw"
            src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DSoda%2BCan&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
          />
        </div>
        <div
          class="flex justify-center"
        >
          <span
            class="py-1 px-2 text-sm bg-white"
          >
            001
          </span>
          <span
            class="py-1 px-2 text-sm bg-black text-white"
          >
            1.25
          </span>
        </div>
      </div>
      <div
        class="flex flex-col"
        data-testid="product"
      >
        <div
          class="relative text-center grow pt-[50%]"
        >
          <img
            alt="Candy Bar"
            data-nimg="fill"
            decoding="async"
            sizes="(max-width: 640px) 100vw, (max-width: 769px) 50vw, (max-width: 1200px) 25vw"
            src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F600%2F300%3Ftext%3DCandy%2BBar&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
          />
          <div
            class="absolute left-1/2 top-1/2 bg-white shadow -translate-x-1/2 -translate-y-1/2 text-nowrap px-2 py-1 text-sm"
          >
            Out of stock
          </div>
        </div>
        <div
          class="flex justify-center"
        >
          <span
            class="py-1 px-2 text-sm bg-white"
          >
            002
          </span>
          <span
            class="py-1 px-2 text-sm bg-black text-white"
          >
            0.75
          </span>
        </div>
      </div>
    </div>
    <div
      class="w-[10rem] max-w-[10rem]"
    >
      <div
        class="bg-black rounded p-4 text-right text-white mb-4"
        data-testid="display"
      >
        <p
          class="text-4xl text-sky-600 font-bold mb-2"
        >
          0
        </p>
        <p
          class="h-6 text-2xl max-w-40 text-ellipsis overflow-hidden mb-2"
        />
        <p
          class="h-10 text-sm"
        />
      </div>
      <div
        class="grid grid-cols-3 gap-2"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.01
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.02
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.05
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.1
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.2
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0.5
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          1
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          2
        </button>
        <button
          aria-label="Insert random coin"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-red-500 text-neutral-50 shadow-sm hover:bg-red-500/90 dark:bg-red-900 dark:text-neutral-50 dark:hover:bg-red-900/90 px-4 py-2 rounded-full w-[3rem] h-[3rem]"
          data-state="closed"
        >
          ?
        </button>
      </div>
      <div
        class="w-full h-[1px] bg-black my-4"
      />
      <div
        class="grid grid-cols-3 gap-2 mb-3"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          1
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          2
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          3
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          4
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          5
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          6
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          7
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          8
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          9
        </button>
        <span
          data-state="closed"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-red-500 text-neutral-50 shadow-sm hover:bg-red-500/90 dark:bg-red-900 dark:text-neutral-50 dark:hover:bg-red-900/90 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
          >
            <div />
          </button>
        </span>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
        >
          0
        </button>
        <span
          data-state="closed"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-neutral-900 text-neutral-50 shadow hover:bg-neutral-900/90 dark:bg-neutral-50 dark:text-neutral-900 dark:hover:bg-neutral-50/90 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
          >
            <div />
          </button>
        </span>
      </div>
      <button
        aria-label="purchase specified product"
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-neutral-900 text-neutral-50 shadow hover:bg-neutral-900/90 dark:bg-neutral-50 dark:text-neutral-900 dark:hover:bg-neutral-50/90 h-9 px-4 py-2 w-full"
      >
        <div />
         Purchase
      </button>
    </div>
  </div>
  <div
    class="w-full h-[1px] bg-black my-4"
  />
  <div
    class="flex gap-2 pt-4 w-full max-w-full flex-wrap"
    data-testid="cash-hopper"
  />
</DocumentFragment>
`;
